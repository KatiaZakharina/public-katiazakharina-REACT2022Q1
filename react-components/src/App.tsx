import { Route, Routes } from 'react-router-dom';

import Layout from 'components/Layout/Layout';
import { AboutUs, BookTour, Home, NotFoundPage } from 'pages';

export default function App() {
  return (
    <Routes>
      <Route path="/" element={<Layout />}>
        <Route index element={<Home />} />
        <Route path="about" element={<AboutUs />} />
        <Route path="book-tour" element={<BookTour />} />
        <Route path="404" element={<NotFoundPage />} />
        <Route path="*" element={<NotFoundPage />} />
      </Route>
    </Routes>
  );
}
