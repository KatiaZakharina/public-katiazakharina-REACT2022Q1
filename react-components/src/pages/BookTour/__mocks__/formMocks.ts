import userEvent from '@testing-library/user-event';

import { fields } from '../Form/Form';
import { TourFormData } from 'features/requests/types';

type InputsEls = { [key in keyof TourFormData]: HTMLInputElement };

export const fakeImage = [new File(['image'], 'image.png', { type: 'image/png' })];

export function getInputs(container: HTMLElement): InputsEls {
  return fields.reduce((acc, curr) => {
    const input = container.querySelector(`[name=${curr}]`) as HTMLInputElement;

    if (input) {
      acc[curr] = input;
    }
    return acc;
  }, {} as InputsEls);
}

export async function fillInputs(data: TourFormData, inputs: InputsEls) {
  userEvent.type(inputs.firstName, data.firstName);
  userEvent.type(inputs.lastName, data.lastName);
  userEvent.type(inputs.email, data.email);
  userEvent.type(inputs.date, data.date);

  userEvent.selectOptions(inputs.destination, data.destination);

  const file = data.pcr;
  userEvent.upload(inputs.pcr, file);

  if (data.getNotification) userEvent.click(inputs.getNotification);
  if (data.withChildren) userEvent.click(inputs.withChildren);
}
