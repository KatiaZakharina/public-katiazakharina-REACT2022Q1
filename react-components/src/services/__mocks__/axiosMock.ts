import axios from 'axios';

export const mockAxiosGet = (responseData: unknown) => {
  jest.mock('axios');
  const axiosMock = axios;

  axiosMock.create = jest.fn();
  axiosMock.get = jest.fn();

  (axiosMock.create as jest.Mock).mockReturnValue(axiosMock);
  (axiosMock.get as jest.Mock).mockResolvedValue(responseData);

  return axiosMock;
};
