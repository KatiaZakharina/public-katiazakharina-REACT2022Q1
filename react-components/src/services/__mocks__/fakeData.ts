export const locationSearchAPIData = {
  suggestions: [
    {
      entities: [
        {
          destinationId: '504261',
        },
      ],
    },
  ],
};

export const hotels = [
  {
    id: 455577,
    name: 'Motel 6 Galveston, TX',
    starRating: 2,
    address: {
      locality: 'Galveston',
      countryName: 'United States',
    },
    landmarks: [
      {
        label: 'City center',
        distance: '4.0 miles',
      },
    ],
    ratePlan: {
      price: {
        exactCurrent: 57.56,
      },
    },
    neighbourhood: 'Galveston',
    optimizedThumbUrls: {
      srpDesktop:
        'https://exp.cdn-hotels.com/hotels/1000000/1000000/996400/996345/e1993493_z.jpg?impolicy=fcrop&w=250&h=140&q=high',
    },
  },
  {
    id: 118521,
    name: 'Beachcomber Inn',
    starRating: 2,
    address: {
      locality: 'Galveston',
      countryName: 'United States',
    },
    landmarks: [
      {
        label: 'City center',
        distance: '3.3 miles',
      },
    ],
    ratePlan: {
      price: {
        exactCurrent: 79.25,
      },
    },
    neighbourhood: 'Galveston',
    optimizedThumbUrls: {
      srpDesktop:
        'https://exp.cdn-hotels.com/hotels/1000000/10000/7700/7682/6ad0ca47_z.jpg?impolicy=fcrop&w=250&h=140&q=high',
    },
  },
];

export const hotelsAPIData = {
  data: {
    body: {
      searchResults: {
        results: hotels,
      },
    },
  },
};

export const hotelDetailsAPIData = {
  result: 'OK',
  data: {
    body: {
      propertyDescription: {
        address: {
          fullAddress: '808 61st St, Galveston, TX, 77551, United States of America',
        },
        name: 'Candlewood Suites Galveston, an IHG Hotel',
        starRating: 2.5,
        featuredPrice: {
          currentPrice: {
            formatted: '$89',
          },
        },
        tagline: ['<b>Hotel with outdoor pool, near Moody Gardens </b>'],
      },
      guestReviews: {
        brands: {
          formattedScale: '10',
          formattedRating: '7.2',
        },
      },
      atAGlance: {
        keyFacts: {
          arrivingLeaving: ['Check-in time 3 PM-10:00 PM', 'Check-out time is  11 AM'],
        },
      },
      amenities: [
        {
          heading: 'In the hotel',
          listItems: [
            {
              heading: 'Food and drink',
              listItems: ['Coffee/tea in a common area', 'Barbecue grills'],
            },
            {
              heading: 'Languages Spoken',
              listItems: ['English', 'Hindi', 'Spanish', 'Urdu'],
            },
          ],
        },
        {
          heading: 'In the room',
          listItems: [
            {
              heading: 'Home comforts',
              listItems: ['Air conditioning', 'Coffee/tea maker', 'Iron/ironing board'],
            },
          ],
        },
      ],
      smallPrint: {
        optionalExtras: [
          '<p><strong>Pets</strong> are allowed for an extra charge of USD 75 per pet, per night</p>Service animals exempt from fees',
        ],
      },
    },
  },
  neighborhood: {
    neighborhoodName: 'Tiki Island',
  },
};

export const hotelImg =
  'https://exp.cdn-hotels.com/hotels/2000000/1990000/1988100/1988040/a32e4d89_{size}.jpg';

export const hotelImgAPIData = {
  hotelImages: [
    {
      baseUrl: hotelImg,
    },
  ],
};
