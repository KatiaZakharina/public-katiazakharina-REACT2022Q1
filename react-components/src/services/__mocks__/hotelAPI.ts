import { DefaultRequestBody, RequestHandler, rest } from 'msw';

import { TourService } from 'services/TourService';
import {
  locationSearchAPIData,
  hotelDetailsAPIData,
  hotelImgAPIData,
  hotelsAPIData,
} from './fakeData';

type Handlers =
  | 'locationIdResponse'
  | 'hotelsResponse'
  | 'hotelDetailsResponse'
  | 'hotelImageResponse';

type HandlersFactorySetup = { handler: Handlers; status: number }[];
type HandlersListType = { [key in Handlers]: { path: string; correctData: unknown } };

export class HandlersFactory {
  public handlers: RequestHandler[];

  handlersList: HandlersListType = {
    locationIdResponse: { path: '/locations/v2/search', correctData: locationSearchAPIData },
    hotelsResponse: { path: '/properties/list', correctData: hotelsAPIData },
    hotelDetailsResponse: { path: '/properties/get-details', correctData: hotelDetailsAPIData },
    hotelImageResponse: { path: '/properties/get-hotel-photos', correctData: hotelImgAPIData },
  };

  basePath = `https://${process.env.REACT_APP_RAPID_API_HOST}`;

  constructor(setup: HandlersFactorySetup) {
    this.handlers = this.setHandlers(setup);
  }

  setHandlers(setup: HandlersFactorySetup): RequestHandler[] {
    const handlers: RequestHandler[] = [];

    setup.forEach((handlerSetup) => {
      const handlerInfo = this.handlersList[handlerSetup.handler];
      const response = handlerSetup.status === 200 ? handlerInfo.correctData : null;

      handlers.push(
        rest.get(`${this.basePath}${handlerInfo.path}`, (req, res, ctx) => {
          return res(ctx.status(handlerSetup.status), ctx.json(response as DefaultRequestBody));
        })
      );
    });

    return handlers;
  }
}
