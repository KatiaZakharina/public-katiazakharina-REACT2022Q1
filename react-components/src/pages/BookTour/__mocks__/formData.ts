import { formatYmd } from 'services/dateFormatter';
import { TourFormData } from '../Form/FormFields';
import { fakeImage } from './formMocks';

export const validData: TourFormData = {
  firstName: 'Ivan',
  lastName: 'Ivanov',
  email: 'ivan@gmail.com',
  date: formatYmd(new Date()),
  destination: 'France',
  withChildren: false,
  pcr: fakeImage,
  getNotification: true,
};

export const invalidData: TourFormData = {
  firstName: 'Iva124',
  lastName: '^&8f#',
  email: 'ivan@gmail.com',
  date: '2022-04-06',
  destination: '',
  withChildren: false,
  pcr: [],
  getNotification: true,
};
