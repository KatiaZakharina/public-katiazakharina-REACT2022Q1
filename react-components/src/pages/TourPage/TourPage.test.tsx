import '@testing-library/jest-dom';
import { screen, waitForElementToBeRemoved } from '@testing-library/react';
import { setupServer } from 'msw/node';

import { HandlersFactory } from 'services/__mocks__/hotelAPI';
import { customRender } from 'test/__mocks__/customRender';
import App from 'App';

describe('TourPage', () => {
  describe('with correct API response', () => {
    const correctHandlers = new HandlersFactory([
      { handler: 'hotelDetailsResponse', status: 200 },
      { handler: 'hotelImageResponse', status: 200 },
    ]).handlers;
    const server = setupServer(...correctHandlers);

    beforeAll(() => server.listen());
    afterEach(() => server.resetHandlers());
    afterAll(() => server.close());

    it('displays hotel details', async () => {
      const tourId = '12345';
      const { asFragment } = customRender(<App />, `/tours/${tourId}`);
      expect(screen.getByTestId('tour_details')).toBeInTheDocument();

      await waitForElementToBeRemoved(() => screen.queryByTestId('preloader'));

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with error details API response', () => {
    const consoleErrorFn = jest.spyOn(console, 'error').mockImplementation(() => jest.fn());

    const errorCode = 500;
    const emptyHandlers = new HandlersFactory([
      { handler: 'hotelDetailsResponse', status: errorCode },
      { handler: 'hotelImageResponse', status: errorCode },
    ]).handlers;

    const server = setupServer(...emptyHandlers);

    beforeAll(() => server.listen());
    afterEach(() => server.resetHandlers());
    afterAll(() => {
      server.close();
      consoleErrorFn.mockRestore();
    });

    it('displays error message', async () => {
      const tourId = '12345';
      const { asFragment } = customRender(<App />, `/tours/${tourId}`);
      expect(screen.getByTestId('tour_details')).toBeInTheDocument();

      await waitForElementToBeRemoved(() => screen.queryByTestId('preloader'));

      expect(asFragment()).toMatchSnapshot();
      expect(consoleErrorFn).toHaveBeenCalled();
    });
  });
});
