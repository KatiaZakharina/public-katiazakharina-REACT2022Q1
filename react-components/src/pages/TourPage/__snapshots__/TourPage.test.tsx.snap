// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TourPage with correct API response displays hotel details 1`] = `
<DocumentFragment>
  <header
    class="sc-bdvvtL kSlYyk"
  >
    <section
      class="sc-eCImPb dskChN"
    >
      <a
        class=""
        href="/"
      >
        <img
          alt="logo image"
          class="sc-hKwDye eYBWIV"
          src="logo.svg"
        />
      </a>
      <nav
        class="sc-gsDKAQ bAkjkz"
      >
        <a
          class="sc-dkPtRN cMHRyC"
          href="/about"
        >
          About us
        </a>
        <a
          class="sc-dkPtRN cMHRyC"
          href="/book-tour"
        >
          Book tour
        </a>
        <a
          class="sc-dkPtRN cMHRyC"
          href="/404"
        >
          404
        </a>
      </nav>
    </section>
  </header>
  <div
    class="sc-GEbAx bnyQPd"
    data-testid="tour_details"
  >
    <section
      class="sc-eCImPb sc-fWCJzd dskChN kwJKmZ"
      data-testid="tour_details"
    >
      <div
        class="sc-clIzBv briFuw"
      >
        <button
          class="sc-Galmp exLVkH"
        >
          ‚Üê Back
        </button>
        <a
          class="sc-faUpoM bcbDck"
          href="/"
        >
          Tours
        </a>
         / 
        <a
          class="sc-faUpoM bcbDck"
          href="/tours/12345"
        >
          Candlewood Suites Galvest...
        </a>
         / 
      </div>
      <section
        class="sc-dvQaRk cGyGES"
      >
        <div
          class="sc-jIkXHa bKella"
        >
          <h1
            class="sc-jOxtWs gCUsFv"
          >
            Candlewood Suites Galveston, an IHG Hotel
          </h1>
          <span
            class="sc-hmjpVf dcEkvA"
          >
            <b>
              Hotel with outdoor pool, near Moody Gardens 
            </b>
          </span>
        </div>
        <p
          class="sc-eLwHnm iqOKBU"
        >
          808 61st St, Galveston, TX, 77551, United States of America
        </p>
        <div
          class="sc-ZOtfp erSvxk"
        >
          <img
            class="sc-bTfYFJ jTxfRF"
            src="https://exp.cdn-hotels.com/hotels/2000000/1990000/1988100/1988040/a32e4d89_z.jpg"
          />
          <div
            class="sc-kHOZwM bfeloR"
          >
            <ul
              class="sc-giYglK bkJEMY"
            >
              <li
                class="sc-ezbkAF cDSUtC"
              >
                <img
                  class="sc-bYoBSM dTBcIh"
                  src="[object Object]"
                />
                2.5 stars
              </li>
              <li
                class="sc-ezbkAF cDSUtC"
              >
                <img
                  class="sc-bYoBSM dTBcIh"
                  src="[object Object]"
                />
                $89 
              </li>
              <li
                class="sc-ezbkAF cDSUtC"
              >
                <img
                  class="sc-bYoBSM dTBcIh"
                  src="[object Object]"
                />
                7.2/10 
              </li>
              <li
                class="sc-ezbkAF cDSUtC"
              >
                <img
                  class="sc-bYoBSM dTBcIh"
                  src="[object Object]"
                />
                Tiki Island neighborhood
              </li>
              <li
                class="sc-ezbkAF cDSUtC"
              >
                <img
                  class="sc-bYoBSM dTBcIh"
                  src="[object Object]"
                />
                Check-in time 3 PM-10:00 PM 
              </li>
              <li
                class="sc-ezbkAF cDSUtC"
              >
                <img
                  class="sc-bYoBSM dTBcIh"
                  src="[object Object]"
                />
                Check-out time is  11 AM 
              </li>
            </ul>
          </div>
        </div>
      </section>
      <section
        class="sc-dvQaRk cGyGES"
      >
        <h2
          class="sc-TBWPX cPcIzM"
        >
          Amenities
        </h2>
        <div
          class="sc-jUosCB fcYQXD"
        >
          <div>
            <h3
              class="sc-jQrDum bpwzsP"
            >
              In the hotel
            </h3>
            <ul
              class="sc-fvxzrP eSSaBh"
            >
              <div>
                <h3
                  class="sc-jQrDum bpwzsP"
                >
                  Food and drink
                </h3>
                <ul
                  class="sc-fvxzrP eSSaBh"
                >
                  <li
                    class="sc-fbyfCU hcERXm"
                  >
                    Coffee/tea in a common area
                  </li>
                  <li
                    class="sc-fbyfCU hcERXm"
                  >
                    Barbecue grills
                  </li>
                </ul>
              </div>
              <div>
                <h3
                  class="sc-jQrDum bpwzsP"
                >
                  Languages Spoken
                </h3>
                <ul
                  class="sc-fvxzrP eSSaBh"
                >
                  <li
                    class="sc-fbyfCU hcERXm"
                  >
                    English
                  </li>
                  <li
                    class="sc-fbyfCU hcERXm"
                  >
                    Hindi
                  </li>
                  <li
                    class="sc-fbyfCU hcERXm"
                  >
                    Spanish
                  </li>
                  <li
                    class="sc-fbyfCU hcERXm"
                  >
                    Urdu
                  </li>
                </ul>
              </div>
            </ul>
          </div>
          <div>
            <h3
              class="sc-jQrDum bpwzsP"
            >
              In the room
            </h3>
            <ul
              class="sc-fvxzrP eSSaBh"
            >
              <div>
                <h3
                  class="sc-jQrDum bpwzsP"
                >
                  Home comforts
                </h3>
                <ul
                  class="sc-fvxzrP eSSaBh"
                >
                  <li
                    class="sc-fbyfCU hcERXm"
                  >
                    Air conditioning
                  </li>
                  <li
                    class="sc-fbyfCU hcERXm"
                  >
                    Coffee/tea maker
                  </li>
                  <li
                    class="sc-fbyfCU hcERXm"
                  >
                    Iron/ironing board
                  </li>
                </ul>
              </div>
            </ul>
          </div>
        </div>
      </section>
      <section
        class="sc-dvQaRk sc-dtMgUX cGyGES jgUVxI"
      >
        <p
          class="sc-cZMNgc dsTFMK"
        />
        <p>
          <strong>
            Pets
          </strong>
           are allowed for an extra charge of USD 75 per pet, per night
        </p>
        Service animals exempt from fees
        <p />
      </section>
    </section>
  </div>
  <footer
    class="sc-gKclnd MQsgS"
  >
    <section
      class="sc-eCImPb dskChN"
    >
      <a
        href="https://rs.school/react/"
        rel="noreferrer"
        target="_blank"
      >
        <img
          class="sc-iCfMLu cYkerD"
          src="rs_school_js.svg"
        />
      </a>
      <div
        class="sc-furwcr eVEnJl"
      >
        <a
          href="https://github.com/KatiaZakharina"
          rel="noreferrer"
          target="_blank"
        >
          Katia Zakharina
        </a>
        <span>
          2022
        </span>
      </div>
    </section>
  </footer>
</DocumentFragment>
`;

exports[`TourPage with error details API response displays error message 1`] = `
<DocumentFragment>
  <header
    class="sc-bdvvtL kSlYyk"
  >
    <section
      class="sc-eCImPb dskChN"
    >
      <a
        class=""
        href="/"
      >
        <img
          alt="logo image"
          class="sc-hKwDye eYBWIV"
          src="logo.svg"
        />
      </a>
      <nav
        class="sc-gsDKAQ bAkjkz"
      >
        <a
          class="sc-dkPtRN cMHRyC"
          href="/about"
        >
          About us
        </a>
        <a
          class="sc-dkPtRN cMHRyC"
          href="/book-tour"
        >
          Book tour
        </a>
        <a
          class="sc-dkPtRN cMHRyC"
          href="/404"
        >
          404
        </a>
      </nav>
    </section>
  </header>
  <div
    class="sc-GEbAx bnyQPd"
    data-testid="tour_details"
  >
    <div
      class="sc-kfPuZi hiYQdM"
    >
      <img
        alt="error illustration"
        class="sc-fKVqWL kTIdEB"
        src="no_connection.svg"
      />
      <div
        class="sc-bBHxTw hgVDmX"
      >
        <h2
          class="sc-iwjdpV irfMDu"
        >
          Internal Server Error
        </h2>
        <p
          class="sc-cxpSdN jljmov"
        >
          Sorry, there were some technical issues while processing yous request
        </p>
        <p
          class="sc-llYSUQ itJvmD"
        >
          ERROR CODE: 500
        </p>
      </div>
    </div>
  </div>
  <footer
    class="sc-gKclnd MQsgS"
  >
    <section
      class="sc-eCImPb dskChN"
    >
      <a
        href="https://rs.school/react/"
        rel="noreferrer"
        target="_blank"
      >
        <img
          class="sc-iCfMLu cYkerD"
          src="rs_school_js.svg"
        />
      </a>
      <div
        class="sc-furwcr eVEnJl"
      >
        <a
          href="https://github.com/KatiaZakharina"
          rel="noreferrer"
          target="_blank"
        >
          Katia Zakharina
        </a>
        <span>
          2022
        </span>
      </div>
    </section>
  </footer>
</DocumentFragment>
`;
